full-node/sov-stf-runner/src/prover_service/blocking_prover.rs

```diff
@@ -0,0 +1,126 @@
+use crate::{ProofGenConfig, Prover, RollupProverConfig, StateTransitionData};
+
+/// Prover that blocks the current thread and creates a ZKP proof.
+pub struct BlockingProver<StateRoot, Witness, Da, Vm, V>
+where
+    StateRoot: Serialize + DeserializeOwned + Clone + AsRef<[u8]>,
+    Witness: Serialize + DeserializeOwned,
+    Da: DaService,
+    Vm: ZkvmHost,
+    V: StateTransitionFunction<Vm::Guest, Da::Spec> + Send + Sync,
+{
+    prover: Mutex<Option<Prover<V, Da, Vm>>>,
+    #[allow(clippy::type_complexity)]
+    witness: Mutex<HashMap<Hash, StateTransitionData<StateRoot, Witness, Da::Spec>>>,
+    zk_storage: V::PreState,
+}
+
+impl<StateRoot, Witness, Da, Vm, V> BlockingProver<StateRoot, Witness, Da, Vm, V>
+where
+    StateRoot: Serialize + DeserializeOwned + Clone + AsRef<[u8]> + Send + Sync,
+    Witness: Serialize + DeserializeOwned + Send + Sync,
+    Da: DaService,
+    Vm: ZkvmHost,
+    V: StateTransitionFunction<Vm::Guest, Da::Spec> + Send + Sync,
+    V::PreState: Clone + Send + Sync,
+{
+    /// Creates a new prover.
+    pub fn new(
+        vm: Vm,
+        zk_stf: V,
+        da_verifier: Da::Verifier,
+        config: Option<RollupProverConfig>,
+        zk_storage: V::PreState,
+    ) -> Self {
+        let prover = config.map(|config| {
+            let stf_verifier =
+                StateTransitionVerifier::<V, Da::Verifier, Vm::Guest>::new(zk_stf, da_verifier);
+
+            let config: ProofGenConfig<V, Da, Vm> = match config {
+                RollupProverConfig::Simulate => ProofGenConfig::Simulate(stf_verifier),
+                RollupProverConfig::Execute => ProofGenConfig::Execute,
+                RollupProverConfig::Prove => ProofGenConfig::Prover,
+            };
+
+            Prover { vm, config }
+        });
+
+        Self {
+            prover: Mutex::new(prover),
+            witness: Mutex::new(HashMap::new()),
+            zk_storage,
+        }
+    }
+}
+
+#[async_trait]
+impl<StateRoot, Witness, Da, Vm, V> ProverService for BlockingProver<StateRoot, Witness, Da, Vm, V>
+where
+    StateRoot: Serialize + DeserializeOwned + Clone + AsRef<[u8]> + Send + Sync,
+    Witness: Serialize + DeserializeOwned + Send + Sync,
+    Da: DaService,
+    Vm: ZkvmHost,
+    V: StateTransitionFunction<Vm::Guest, Da::Spec> + Send + Sync,
+    V::PreState: Clone + Send + Sync,
+{
+    type StateRoot = StateRoot;
+
+    type Witness = Witness;
+
+    type DaService = Da;
+
+    async fn submit_witness(
+        &self,
+        state_transition_data: StateTransitionData<
+            Self::StateRoot,
+            Self::Witness,
+            <Self::DaService as DaService>::Spec,
+        >,
+    ) {
+        let header_hash = state_transition_data.da_block_header.hash().into();
+        self.witness
+            .lock()
+            .expect("Lock was poisoned")
+            .insert(header_hash, state_transition_data);
+    }
+
+    async fn prove(&self, block_header_hash: Hash) -> Result<(), ProverServiceError> {
```

Comment:
I see that `ZkStorage` is an attribute of prover.

It makes an assumption, that ZkStorage is stateless, so it is free to clone it. But this is implicit assumption.

What are the cons of passing zk_storage as parameter to `prove` method? And RollupBluePrint creates it as before?

Reply:
Ok, it was already like that in rollup runner , so no need to resolve it this PR.

But I still would like to have your comment. If you think that this implicit assumption might bring some problems along the way, we need to address it in separate issue.

Reply:
There are several reasons for this design:

1. In the above code, the `ProverService` owns the storage and (currently) is not cloned anywhere. By passing it explicitly, we leak it, and the `ZkStorage` would be accessible in more places unnecessarily.
2. Once we implement `ProverService` for bonsai, we won't need it at all, so it should not be exposed to the rest of the system.
3. In the next "multithreaded" version of the `ProverService`, every thread will need to have its own clone of the `ZkStorage`. If there ever will be a version of stateful `ZkStorage`, then the implementer needs to make sure that the `Clone` is implemented correctly and can be used this way.

Also: The assumption about being stateless is kind of explicit, as the `set/get` methods accept a `witness` and don't have `&mut self`.

Reply:
1. I agree that leaking one more associated type is not nice.
2. I don't fully get it, I thought that Bonsai will be just running guest VM, which still needs ZkStorage
3. This part, where I don't feel comfortable. Have correct clone behavior is not the same as being stateless. For instance, having `Arc<RwLock<Vec>>` will have correct clone, but underlying data still be shared. Of course type system gives a hint here. Same goes with "not having `&mut self`". ProverStorage is mutable, and follow the same pattern.

I agree that it is not a problem at the moment, as ZkStorage indeed stateless, but as this assumption leaks into other parts of the system, I would prefer not to pull it under the rug.

If you are sure, that this approach won't back fire in the future, I trust you.